### DISCUSSION TITLE
Basics batch syntax for newbie

### DISCUSSION THREAD

**Discussion by VadeInPace** [ImageMagick/ImageMagick#6404] [Help] [answered]
### ImageMagick version

7.1.1-4 Q16-HDRI

### Operating system, version and so on

Linux

### Description

Hi there,

Even after lot of reading, I can't figure out how batch syntax is working.
I want to convert pdf file to png file.
pdf file name are with space. ie : 
first file.pdf
second file.pdf
I want the output in another directory.
I wrote
```
outdir="~/pngfiles"

for file in *.pdf
do
echo "$file"
   convert \
      -density 600 \
      "$file" \
      -quality 90 \
   "$outdir/"${file%.*}".png"
echo "$outdir/${file%.*}.png"
echo
done
```

But It's not working as expected; File name are split.

I think there is a way without the for loop. I am open to try if this is a better way.
Thanks

**Comment by fmw42** [comment, marked as answer]
I am not a Windows expert, but too many quotes. 

Change

`"$outdir/"${file%.*}".png"`

to

`"$outdir/${file%.*}.png"`



However, you may be able to do it all using mogrify

Create a new directory or be sure one exists where you want the output.  Imagemagick will not create one.

```
cd current_directory
mogrify -path output_directory -density 600 -format png -quality 90 *.png
```

In this case, I am not sure whether the density will be properly applied before reading the file or afterwards. So this may not  work for PDF file that need the density applied before reading the input.

**Reply by VadeInPace** [reply]
Corrected with **.pdf** at the end of the line instead of .png

**Comment by snibgo** [comment]
@VadeInPace: In your command, the quote marks are stripped by bash, so IM never sees them. You must tell bash to pass the outer quotes to IM, so IM will not split the command at the space within the filename. The inner quotes are not needed. I suggest:
```
\"$outdir/${VAR%.*}.png\"
```
EDIT: As a general rule, I suggest not having spaces in filenames, or directory names. They are more trouble than they are worth.

**Comment by VadeInPace** [comment]
There is a convert error I'd like to remove before going further in the batch script
I changed the pdf file name to First.pdf and tried only one file
```
> convert -density 600 First.pdf -quality 90 ~/pngfiles/out.png
convert: delegate library support not built-in 'First.pdf' (XML) @ warning/profile.c/ValidateXMPProfile/1946.
convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `/home/jp/pngfiles/out.png' @ warning/png.c/MagickPNGWarningHandler/1526.

```
How to remove these errors. For information, there are 2 pages in the pdf. Output are in /pngfiles, out-0.png and out-1.png

**Reply by fmw42** [reply]
Do you have Ghostscript installed?  What do you get from 

`convert -version`

The messages are just a warning that one cannot have a grayscale PNG with color profile. You can ignore that.  The first message is the issue.  Did you get a resulting out.png?

**Reply by VadeInPace** [reply]
Yes, ghostscript version 9.52-150000.164.1
```
> convert -version
Version: ImageMagick 7.1.1-4 Q16-HDRI x86_64 20976 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5) 
Delegates (built-in): ltdl png zlib
Compiler: gcc (7.5)
```
Yes 2 output files.
OK, I'll ignore that

**Reply by fmw42** [reply]
You do not have many delegates and your list does not show gs or gslib or Ghostscript.  So perhaps you did not install it correctly.

What do you get from 

`convert -list configure`

**Reply by VadeInPace** [reply]
```
> convert -list configure

Path: /usr/local/lib64/ImageMagick-7.1.1//config-Q16HDRI/configure.xml

Name                  Value
-------------------------------------------------------------------------------
CC                    gcc
CFLAGS                -I/usr/include/libpng16  -fopenmp -Wall -g -O2 -mtune=amdfam10 -fexceptions -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
CODER_PATH            /usr/local/lib64/ImageMagick-7.1.1/modules-Q16HDRI/coders
CONFIGURE             ./configure  '--enable-hdri' '--with-modules'
CONFIGURE_PATH        /usr/local/etc/ImageMagick-7/
COPYRIGHT             Copyright (C) 1999 ImageMagick Studio LLC
CPPFLAGS              -I/usr/local/include/ImageMagick-7
CXX                   g++
CXXFLAGS              -g -O2 -pthread
DEFS                  -DHAVE_CONFIG_H
DELEGATES             png ps video zlib
DISTCHECK_CONFIG_FLAGS  --disable-deprecated  --with-quantum-depth=16  --with-jemalloc=no  --with-umem=no  --with-autotrace=no  --with-dps=no  --with-fftw=no  --with-flif=no  --with-fpx=no  --with-gslib=no  --with-fontpath=  --with-rsvg=no  --with-perl=no 
DOCUMENTATION_PATH    /usr/local/share/doc/ImageMagick-7
EXEC-PREFIX           /usr/local
EXECUTABLE_PATH       /usr/local/bin
FEATURES              DPC HDRI Cipher OpenMP Modules
FILTER_PATH           /usr/local/lib64/ImageMagick-7.1.1/modules-Q16HDRI/filters
GIT_REVISION          20976
HOST                  x86_64-pc-linux-gnu
INCLUDE_PATH          /usr/local/include/ImageMagick-7
LDFLAGS               -L/usr/local/lib64 
LIB_VERSION           0x711
LIB_VERSION_NUMBER    7,1,1,4
LIBRARY_PATH          /usr/local/lib64/ImageMagick-7.1.1
LIBS                                -lz   -lltdl  -lm    -lpthread  -lgomp
NAME                  ImageMagick
PCFLAGS               -fopenmp -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
PREFIX                /usr/local
QuantumDepth          16
RELEASE_DATE          2023-03-18
SHARE_PATH            /usr/local/share/ImageMagick-7
SHAREARCH_PATH        /usr/local/lib64/ImageMagick-7.1.1/config-Q16HDRI
TARGET_CPU            x86_64
TARGET_OS             linux-gnu
TARGET_VENDOR         pc
VERSION               7.1.1
WEBSITE               https://imagemagick.org

Path: [built-in]

Name                  Value
-------------------------------------------------------------------------------
DELEGATES             ltdl png zlib
FEATURES              Cipher DPC HDRI Modules OpenMP(4.5) 
MAGICK_TEMPORARY_PATH /tmp
NAME                  ImageMagick
QuantumDepth          Q16
```

**Reply by VadeInPace** [reply]
If you think the messages are not error but only warning and the files are fine, I'll stay like that

**Comment by fmw42** [comment]
You delegates do not include gs or gslib for Ghostscript. So if you have Ghostscript installed, it is not installed where Imagemagick can find it.

DELEGATES             png ps video zlib

**Reply by VadeInPace** [reply]
You are right, I built IM 3 months ago with the help of snibgo.
As the pdf aren't coloured, and png output have a good looking, I'll stay like that.
Many thanks

**Comment by fmw42** [comment]
Ask snibgo, if he helped you build your system. Perhaps Windows does not show Ghostscript though it is installed?

**Reply by VadeInPace** [reply]
No!, I use Linux, Opensuse Leap 15.4

**Reply by fmw42** [reply]
My understanding is that you do not have Ghostscript installed in Imagemagick. So I am puzzled how you even got an output. I am unaware that is possible if Ghostscript is not installed. I see no sign that Ghostscript is installed in Imagemagick.

**Reply by fmw42** [reply]
You have Ghostscript disabled as shown at

`DISTCHECK_CONFIG_FLAGS  --disable-deprecated  --with-quantum-depth=16  --with-jemalloc=no  --with-umem=no  --with-autotrace=no  --with-dps=no  --with-fftw=no  --with-flif=no  --with-fpx=no  --with-gslib=no  --with-fontpath=  --with-rsvg=no  --with-perl=no 
`

where it shows  `--with-gslib=no`

**Reply by VadeInPace** [reply]
FYI, I used the script to convert about 30 pdf with 1-3 pages and 1 with ~65 pages. It took a lot of CPU and memory+swap (when 65 pages)
Then I used your command with mogrify same memory usage but a bit less time.
Results are exactly the same, same size (same bytes number)
Same warnings at the end for mogrify

**Comment by fmw42** [comment]
Did convert give the same warnings?  You should be using magick not convert on IM 7.  Also magick mogrify not mogrify.

**Reply by VadeInPace** [reply]
```
cd current_directory
mogrify -path output_directory -density 600 -format png -quality 90 *.png
```
I used the above you posted in first answer.
Could you let me know the exact syntax  with magick mogrify ?

**Reply by fmw42** [reply]
same as above. just change mogrify to magick mogrify

**Reply by VadeInPace** [reply]
Many thanks, just what I wanted to know.

**Comment by fmw42** [comment]
Do you have IM 6 and IM 7 installed?  Perhaps you are using IM 6 convert and not using IM 7 and IM 6 has Ghostscript installed. 

 What do you get from 

`convert -version`

and

`magick -version`

It is not good to have multiple versions of IM installed at the same time.

**Reply by VadeInPace** [reply]
```
> convert -version
Version: ImageMagick 7.1.1-4 Q16-HDRI x86_64 20976 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5) 
Delegates (built-in): ltdl png zlib
Compiler: gcc (7.5)

> magick -version
Version: ImageMagick 7.1.1-4 Q16-HDRI x86_64 20976 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5) 
Delegates (built-in): ltdl png zlib
Compiler: gcc (7.5)

```

**Comment by VadeInPace** [comment]
Thanks you fmw42 for your help.
Just to repeat, the both commands, loop with convert and magick mogrify have the exact same result.
The loop take a bit more time but magick mogrify is straightforward to write and faster.
All the work is done.
